migrations/lib/importer/steps/topics.rb

```diff
@@ -0,0 +1,142 @@
+      SELECT COUNT(*)
+      FROM topics
+           LEFT JOIN mapped.ids mapped_topic
+             ON topics.original_id = mapped_topic.original_id  AND mapped_topic.type = ?1
+      WHERE mapped_topic.original_id IS NULL
+    SQL
+
+    rows_query <<~SQL, MappingType::TOPICS, MappingType::CATEGORIES, MappingType::USERS
+      SELECT topics.*,
+              mapped_category.discourse_id        AS discourse_category_id,
+              mapped_user.discourse_id            AS discourse_user_id,
+              mapped_deleted_by_user.discourse_id AS discourse_deleted_by_user_id
+      FROM topics
+           LEFT JOIN mapped.ids mapped_topic
+             ON topics.original_id = mapped_topic.original_id  AND mapped_topic.type = ?1
+           LEFT JOIN mapped.ids mapped_category
+             ON topics.category_id = mapped_category.original_id  AND mapped_category.type = ?2
+           LEFT JOIN mapped.ids mapped_user
+             ON topics.user_id = mapped_user.original_id  AND mapped_user.type = ?3
+           LEFT JOIN mapped.ids mapped_deleted_by_user
+             ON topics.deleted_by_id = mapped_deleted_by_user.original_id  AND mapped_deleted_by_user.type = ?3
+      WHERE mapped_topic.original_id IS NULL
+    SQL
+
+    private
+
+    def transform_row(row)
+      return nil if row[:archetype] != Archetype.private_message && row[:discourse_category_id].nil?
+
+      if (external_id = row[:external_id])
+        return nil unless process_external_id(external_id)
+      end
+
+      row[:archived] ||= false
+      row[:closed] ||= false
+      row[:visible] = true if row[:visible].nil?
+
+      row[:views] ||= DEFAULT_VIEWS
+
+      row[:category_id] = row[:discourse_category_id] ||
+        (UNCATEGORIZED_ID if row[:archetype] != Archetype.private_message)
+      row[:deleted_by_id] = row[:discourse_deleted_by_user_id]
+      row[:user_id] = row[:discourse_user_id] || SYSTEM_USER_ID
+
+      row[:title] = ensure_valid_title(row[:title])
+      row[:last_post_user_id] ||= row[:user_id]
+      row[:slug] = Slug.for(row[:title])
+      row[:bumped_at] = row[:created_at]
+
+      row[:archetype] = ensure_valid_value(
+        value: row[:archetype],
+        allowed_set: ARCHETYPES,
+        default_value: DEFAULT_ARCHETYPE,
+      ) do |_, default_value|
+        puts "   #{row[:id]}: Topic archetype is invalid, defaulting to #{default_value}"
+      end
+      row[:subtype] = ensure_valid_value(
+        value: row[:subtype],
+        allowed_set: SUBTYPES,
+        default_value: nil,
+      ) if row[:subtype]
+      row[:visibility_reason_id] = ensure_valid_value(
+        value: row[:visibility_reason_id],
+        allowed_set: VISIBILITY_REASONS,
+        default_value: nil,
+      ) if row[:visibility_reason_id]
+
+      super
+    end
+
+    def ensure_valid_title(title)
+      cleaned_title = TextCleaner.clean_title(TextSentinel.title_sentinel(title).text)
+      cleaned_title[0, SiteSetting.max_topic_title_length]
+    end
```

Comment:
Looking at: https://github.com/discourse/discourse/blob/4e417223e5d5c880f0bdad3af696c0f381d68859/app/models/topic.rb#L192-L216 

Feels like the title cleanup here could be more thorough, especially when importing into an existing site.  But it's a hot path, even the `TextCleaner` usage here might be a performance drag. Should we lean toward a more thorough cleanup (including at least a watched words check), or keep it as just a simple length check?

Reply:
I'd restrict this to a length check. It's difficult to automatically resolve issues most of those validations would raise. At least not without making a total mess. And our goal should be to import as much content in its original state as possible. Some checks, like scanning and skipping spam, might be an optional step between converter and importer. Most likely as part of the transformer tool.

Reply:
Alright, sure!